### PR TITLE
Add private mechanism to toggle vmap fallback warnings

### LINKED ISSUE #51144: Batched gradient computation can spam warnings
## 🐛 Bug

Batched gradient computation warns when a batching rule is not implemented. However, warnings during backward execution may not be turned into Python warnings, leading to console spam (#50209)

## To Reproduce

```py
import torch
from torch._vmap_internals import vmap
import warnings
warnings.simplefilter('ignore')

device = 'cuda:0'
x = torch.randn(4, requires_grad=True, device=device)
indices = torch.tensor([0, 1], device=device)
y = x.gather(0, indices)

def vjp(v):
    return torch.autograd.grad(y, x, v, retain_graph=True)

vs = torch.randn(2, *y.shape, device=device)
vmap(vjp)(vs)
vmap(vjp)(vs)
vmap(vjp)(vs)

# This script has the following output, despite the warnings filter
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
```

## Challenges

I don't think we are going to fix #50209 anytime soon. Furthermore, using `TORCH_WARN_ONCE` doesn't solve the problem because all operators that go through the fallback use the same exact vmap warning generation code; `TORCH_WARN_ONCE` would cause the warning to appear only at the first operator that raises the warning.

Any solution must take into account that autograd is multi-threaded.

## Potential solutions

Here are a few different proposals for this, categorized into "short-term" and "long-term". I am not suggesting that we do something in the short-term bucket first.

"Short-term"
- Implement a VMAP_WARN_ONCE that allocates a hash map on a heap. If an operator is not in the hash map, then we print a warning and add it to the map. If it is in the map, then we do not print the warning. (Is adding elements to `std::unordered_map` thread safe?)
- By default, disable the vmap fallback warning. Add a private `_ShowVmapWarnings` context manager that enables the vmap fallback warning. We already have [a mechanism to handle propagating thread-local state through autograd threads](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ThreadLocalState.h)

"long term"
- Have an actual "TORCH_WARN_ONCE_PER_PARAMETERS" that isn't some slow implementation around a hashmap. This is probably physically impossible
- (Leveraging a proposal from @ezyang) Have a "performance warnings context manager". When enabled, PyTorch prints on performance warnings. If we had this, then we would classify vmap fallback warnings as "performance warnings".

## Environment

master

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51218 Add private mechanism to toggle vmap fallback warnings**

Fixes #51144.

Context
=======

Users have complained about warning spam from batched gradient
computation. This warning spam happens because warnings in C++ don't
correctly get turned into Python warnings when those warnings arise from
the autograd engine.

To work around that, this PR adds a mechanism to toggle vmap warnings.
By default, the vmap fallback will not warn when it is invoked. However,
by using `torch._C._debug_only_display_vmap_fallback_warnings(enabled)`,
one can toggle the existence of vmap fallback warnings.

This API is meant to be a private, debug-only API. The goal is to be
able to non-intrusively collect feedback from users to improve
performance on their workloads.

What this PR does
=================

This PR adds an option to toggle vmap warnings. The mechanism is
toggling a bool in ATen's global context.

There are some other minor changes:
- This PR adds a more detailed explanation of performance cliffs to the
autograd.functional.{jacobian, hessian} documentation
- A lot of the vmap tests in `test_vmap.py` rely on the fallback warning
to test the presence of the fallback. In test_vmap, I added a context
manager to toggle on the fallback warning while testing.

Alternatives
============

I listed a number of alternatives in #51144. My favorite one is having a new
"performance warnings mode" (this is currently a WIP by some folks on
the team). This PR is to mitigate the problem of warning spam before
a "performance warnings mode" gets shipped into PyTorch

Concerns
========

I am concerned that we are advertising a private API
(`torch._C._debug_only_display_vmap_fallback_warnings(enabled)`) in the
PyTorch documentation. However, I hope the naming makes it clear to
users that they should not rely on this API (and I don't think they have
any reason to rely on the API).

Test Plan
=========

Added tests in `test_vmap.py` to check:
- by default, the fallback does not warn
- we can toggle whether the fallback warns or not

Differential Revision: [D26126419](https://our.internmc.facebook.com/intern/diff/D26126419)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
accepting to unblock
